JIT pass to insert XNNPACK ops

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34547 Enable threading for XNNPACK ops.
* #34319 Pass to remove prepacking ops.
* **#34048 JIT pass to insert XNNPACK ops**
* #34047 Integrate XNNPACK with custom class for packing weights.

Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

kimishpatel committed:
JIT pass to insert XNNPACK ops

Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -0,0 +1,121 @@
+namespace jit {
+
+namespace {
+std::string getFuncName(Value* func_value) {
```
I remember I have this function in quantization.cpp as well

kimishpatel replied:
Yes, need to move this to the common file too.

jerryzh168 commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -0,0 +1,121 @@
+  std::string xnnpack_pattern_before_inline = R"(
+    graph(%linear, %input, %weight, %bias):
+        %packed_weight_bias = xnnpack::linear_prepack(%weight, %bias)
+        %res = xnnpack::linear_packed(%input, %packed_weight_bias)
```
why is this called `xnnpack::linear_packed`? do you have another function called `xnnpack::linear`?

kimishpatel replied:
This is how it was called in the original diff that landed xnnpack (https://github.com/pytorch/pytorch/pull/32509), so kept the naming convention.

dreiss replied:
It's to emphasize that the arguments are not normal weights and biases.

jerryzh168 commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -0,0 +1,121 @@
+
+void insertXNNPACKOps(std::shared_ptr<Graph>& graph) {
+  ConstantPooling(graph);
+  ConstantPropagation(graph);
```
also, we plan to remove these from our code, it's better not to depend on them, could you check if removing them also works?

kimishpatel replied:
Do you mean constant propagation? To replace aten::convolution with conv2d you need this. Else it does not work.

jerryzh168 replied:
yeah, we plan to make our patterns not rely on constant prop.

Do you mean fuse linear? replace aten::_convolution should work without constant prop

kimishpatel replied:
I remember it did not work, because you need to check the parameters values. Some values pass through prim::ListConstruct and without const prop. you do not see the values.
For example output_padding value. This is the one it was not able to find without const prop.

jerryzh168 commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -0,0 +1,121 @@
+        %bias: Tensor? = prim::Constant()
+        %packed_weight_bias = xnnpack::linear_prepack(%weight, %bias)
+        %res = xnnpack::linear_packed(%input, %packed_weight_bias)
+        return (%res))";
```
This is probably not needed

kimishpatel replied:
Why not?

jerryzh168 replied:
this is the same as previous pattern

kimishpatel replied:
Previous pattern is supposed to catch bias=None as well? I will try it out.

jerryzh168 commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -0,0 +1,121 @@
+
+  SubgraphRewriter linear_rewriter_wo_bias;
+  linear_rewriter_wo_bias.RegisterRewritePattern(linear_pattern_wo_bias, xnnpack_pattern_wo_bias);
+  linear_rewriter_wo_bias.runOnGraph(graph);
```
This pattern is not needed I think

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

dreiss commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -0,0 +1,108 @@
+void insertXNNPACKLinearOp(std::shared_ptr<Graph>& graph) {
+  std::string linear_before_inline = R"(
+    graph(%linear, %input, %weight, %bias):
+        %r = prim::CallFunction(%linear, %input, %weight, %bias)
```
This one makes me a bit nervous.  Will this match any 4-arg function called "linear", even if it's not aten::linear or torch.nn.functional.linear?

kimishpatel replied:
This is a good point. Your observation I think is correct. Let me think about this as to what are the caveats here. Tagging @jerryzh168, to clarify something I might have missed.

kimishpatel replied:
So this is doable. I think we can do more stringent checking. I will do it outside of this stack, since there are a few common functions we share with quantization and if we make a change there, it affects both. And I will have to make changes in quantization.cpp as well.

dreiss commented:
Oh, as discussed in the other PR, let's change the namespace to _xnnpack.

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

jerryzh168 commented:
You need to filter the function name, see quantization.cpp for examples

Sent from my iPhone

On Mar 11, 2020, at 10:27, Kimish Patel <notifications@github.com> wrote:

﻿

@kimishpatel commented on this pull request.

________________________________

In torch/csrc/jit/passes/xnnpack_rewrite.cpp<https://github.com/pytorch/pytorch/pull/34048#discussion_r391139568>:

> +#include <torch/csrc/jit/passes/fuse_linear.h>
+#include <torch/csrc/jit/passes/graph_rewrite_helper.h>
+#include <torch/csrc/jit/passes/subgraph_rewrite.h>
+#include <torch/csrc/jit/ir/subgraph_matcher.h>
+
+namespace torch {
+namespace jit {
+
+#ifdef USE_XNNPACK
+
+namespace {
+
+void insertXNNPACKLinearOp(std::shared_ptr<Graph>& graph) {
+  std::string linear_before_inline = R"(
+    graph(%linear, %input, %weight, %bias):
+        %r = prim::CallFunction(%linear, %input, %weight, %bias)


This is a good point. Your observation I think is correct. Let me think about this as to what are the caveats here. Tagging @jerryzh168<https://github.com/jerryzh168>, to clarify something I might have missed.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/34048#discussion_r391139568>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABF2R2NQAY34UTSGX4BLC3DRG7CXFANCNFSM4K7IUZUQ>.

kimishpatel commented:
> You need to filter the function name, see quantization.cpp for examples

Thats what it does. David's point is how reliable it is to just rely on function name for filtering, if I understood his point correctly.

jerryzh168 commented:
Yeah it will match all 4-arg functions called linear, maybe we can match the full namespace, just need to change getFuncName to return full qualified name

Sent from my iPhone

On Mar 11, 2020, at 11:48, Kimish Patel <notifications@github.com> wrote:

﻿

You need to filter the function name, see quantization.cpp for examples

Thats what it does. David's point is how reliable it is to just rely on function name for filtering, if I understood his point correctly.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/34048#issuecomment-597805384>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABF2R2I7QIKJ3HSIOJS7IGLRG7MHHANCNFSM4K7IUZUQ>.

kimishpatel commented:
> Yeah it will match all 4-arg functions called linear, maybe we can match the full namespace, just need to change getFuncName to return full qualified name
> […](#)
> Sent from my iPhone On Mar 11, 2020, at 11:48, Kimish Patel <notifications@github.com> wrote: ﻿ You need to filter the function name, see quantization.cpp for examples Thats what it does. David's point is how reliable it is to just rely on function name for filtering, if I understood his point correctly. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub<[#34048 (comment)](https://github.com/pytorch/pytorch/pull/34048#issuecomment-597805384)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABF2R2I7QIKJ3HSIOJS7IGLRG7MHHANCNFSM4K7IUZUQ>.

Yes, I think that would be good.

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

kimishpatel commented:
@dreiss, addressed your comments.

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass to insert XNNPACK ops"


Summary:
Rewrites the graph to insert xnnpack prepack and packed run ops for
conv2d and linear.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185658](https://our.internmc.facebook.com/intern/diff/D20185658)

[ghstack-poisoned]
